_posts/2024-12-04-enable-pluggable-storage-in-opensearch-vectordb.md

```diff
@@ -0,0 +1,255 @@
+          "price": 12.2
+        }
+      },
+      {
+        "_index": "knn-index",
+        "_id": "2",
+        "_score": 0.6666667,
+        "_source": {
+          "my_vector": [
+            2.5,
+            3.5
+          ],
+          "price": 7.1
+        }
+      }
+    ]
+  }
+}
+```
+
+### Step 4: Take a snapshot
+
+Take a snapshot of the index. For detailed steps, see [Take and restore snapshots](https://opensearch.org/docs/latest/tuning-your-cluster/availability-and-recovery/snapshots/snapshot-restore/). After taking the snapshot, delete the `knn-index` so it is no longer available locally.
+
+### Step 5: Create a searchable snapshot index from the snapshot
+
+Use the following request to restore the original index from the snapshot as a searchable snapshot:
+
+```json
+POST _snapshot/<SNAPSHOT_REPO>/<SNAPSHOT_NAME>/_restore
+{
+  "storage_type": "remote_snapshot",
+  "indices": "knn-index"
+}
+```
+
+To verify the successful creation of the searchable snapshot index, use the following request:
+
+```json
+GET /_cat/indices
+```
+For more information, see [Create a searchable snapshot index](https://opensearch.org/docs/latest/tuning-your-cluster/availability-and-recovery/snapshots/searchable_snapshot/#create-a-searchable-snapshot-index).
+
+### Step 6: Run a vector search query
+
+Run a vector search query on the searchable snapshot index:
+
+```json
+POST knn-index/_search
+{
+  "query": {
+    "knn": {
+      "my_vector": {
+        "vector": [2, 3],
+        "k": 2
+      }
+    }
+  }
+}
+```
+
+The query returns the same results as the query on the local index in [Step 3](#step-3-query-the-local-index).
+
+## Conclusion
+
+By introducing an abstract I/O layer that uses Lucene's **Directory** abstraction, we eliminated the native engines' dependency on file-based APIs that limited storage to local file systems. This change allows the vector engine to read graph data structures from any storage system supported by OpenSearch's **Directory** implementation. Our extensive benchmark testing confirmed that this change maintained the original file API-based approach's search performance. Notably, we observed no regression in search times after the graphs were loaded into memory (graph loading is a one-time operation for a properly scaled cluster).
+
+With this new read interface, you can now use any **Directory** implementation in OpenSearch with vector indexes. This added flexibility makes it possible to store vector data in remote storage solutions like S3.
+
+## Next steps
+
+In the 2.18 release, we introduced the capability to enable vector search queries using Lucene's **Directory** and **IndexInput** classes. Looking ahead, the 2.19 release will expand this functionality to the native index creation process. Specifically, the k-NN plugin will start using the **IndexOutput** class to write graph files directly to segments. For more information, see [this GitHub issue](https://github.com/opensearch-project/k-NN/issues/2033).
+
```

Comment:
Line 253, first sentence: "the ability to use vector search queries with Lucene's..."?